packit/upstream.py

```diff
@@ -539,11 +539,7 @@ def _fix_spec_prep(self, version):
             new_setup_line += args_match.group(2)
         else:
             new_setup_line += m[2]
-        if not self.package_config.upstream_package_name:
-            raise PackitException(
-                '"upstream_package_name" is not set: unable to fix the spec file; please set it.'
-            )
-        new_setup_line += f" -n {self.package_config.upstream_package_name}-{version}"
+        new_setup_line += f" -n {archive}"
```

Comment:
I'm sorry to say this but this is not what we want.

`-n` accepts the prefix within the tarball, not the tarball path:
```
$ tar -tf packitos-0.9.1.dev2+gd4f6f93.tar.gz
packitos-0.9.1.dev2+gd4f6f93/
packitos-0.9.1.dev2+gd4f6f93/.zuul.yaml
packitos-0.9.1.dev2+gd4f6f93/CHANGELOG.md
packitos-0.9.1.dev2+gd4f6f93/CONTRIBUTING.md
packitos-0.9.1.dev2+gd4f6f93/DCO
```

in this case, correct argument to `-n` is `packitos-0.9.1.dev2+gd4f6f93`


I'm assuming the issue is about being able to specify `self.package_config.upstream_package_name` - I'd suggest creating a new config option, "archive_prefix" for this so people can easily specify it.

@lachmanfrantisek WDYT?

Reply:
> as for patching run_command - would it be possible not to patch, run the actual command thus making the test integration?

This is unit test so I think it is ok as it is. I will add also integration test

Reply:
> @lachmanfrantisek WDYT?

It's really tricky that there are double suffixes.

- We can try to remove suffix manually (well-know double suffixes can be handled separately).
  - The manual approach is not so nice but the variability is not so wide.
- What about getting the correct name from the archive content?
  - It's dependent on the archive type, but at least for tars, [we can do it](https://docs.python.org/3/library/tarfile.html)).
- Specifying `archive_extension`.
  - Is that what you mean by `archive_prefix`?
  - Easy for us, but a bit lazy -- I would try to avoid having another config option if possible.

Reply:
yes, this is tricky as hell

> We can try to remove suffix manually

I don't follow. This is the most common way of creating upstream release tarballs: have a single dir inside: `"{project_name}-{version}"`

> What about getting the correct name from the archive content? 

yes, we could do that, sounds like the ideal default

> Is that what you mean by archive_prefix?

No. When we change the %setup line, we need to make sure that -n has the correct argument so rpmbuild can unpack the archive correctly.

Does this answer your questions?

Reply:
> I don't follow. This is the most common way of creating upstream release tarballs: have a single dir inside: `"{project_name}-{version}"`

- We need to set the directory inside an archive.
- My statement: The name of the directory is the same as the name of the archive in most cases.
- If we remove the suffix from the archive, we get the correct name.

This will solve the problem that they can use both upstream and downstream name for the tarball. (Aha, and that's what you proposed with the `archive_prefix`.)

>> What about getting the correct name from the archive content?
> yes, we could do that, sounds like the ideal default

:+1:

Reply:
> > I don't follow. This is the most common way of creating upstream release tarballs: have a single dir inside: `"{project_name}-{version}"`

This is correct when we create the archive. But my assumption here is that the user is actually creating the archive while we are updating the %setup line

>     * We need to set the directory inside an archive.
> 
>     * My statement: The name of the directory is the same as the name of the archive in most cases.
> 
>     * If we remove the suffix from the archive, we get the correct name.

Yes and no: we'd start doing something which is not common - I'd expect some of our users to complain (but they complain all the time :D :D ).

I'd personally prefer to stick to the common workflow and diverge only when it's causing us big troubles.

Reply:
OK, I'm a bit confused about the outcode and will try to summarize it:

- What we need on that line is the name of the directory in the archive.
- Usually, it's a combination of the (upstream or downstream) name and the version.
    - By now, we are using the upstream name but some projects use the downstream name.
- Possible solutions:
  - Get the directory name for known types.
    - Types we can easily implement that for.
    - Can be combined with any of the following options.
  - Support configuration of the `archive_prefix` (or what about `source_prefix`?) which we can use there
    (`f" -n {self.package_config.source_prefix if self.package_config.source_prefix else self.package_config.upstream_package_name}-{version}"`)
  - Try to use an archive name (filename without extension(s)).
    - Can be tricky because of the double extensions.
    - It's not a rule that the archive name is equal to that directory.


- Is that correct?
- Any other option?

Reply:
@lachmanfrantisek @TomasTomecek 

Please can you summarize this? What is the expected approach? Should I implement multiple techniques as a fallback option if primary will fail?


btw. #959 will fix the issue #868 based on which this PR was created

Reply:
I tried to summarize that in the above comment and waiting for the confirmation by Tomáš.

> Should I implement multiple techniques as a fallback option if primary will fail?

The thing is that we can't easily test if it works or not.

> btw. #959 will fix the issue #868 based on which this PR was created

It will solve the problem for jkonecny, but not this issue.

Reply:
@lachmanfrantisek sorry I had old window opened and didn't saw your comment

Reply:
I vote for extracting dir from the archive and have the possibility to configure it in case of need.

regarding naming -  I will avoid using 'prefix' in the name because I see it as counterintuitive, I will prefer something like dir_inside_archive

Reply:
> dir_inside_archive

But it's not the full name of that directory. It's really a prefix of it. Or we can use something like `source_name`.

Reply:
I had a call with @lachmanfrantisek and we agreed on the following approach:
1. we will try to extract the directory form archive
2. if an archive type will be not supported by packit we will check if an option inner_archive_dir_template(is there a better name for it?) is configured
3. we will try to use the current approach -  `{self.package_config.upstream_package_name}-{version}`

Note:
step 2 and 3 will be merged - 2 will be the default template if implementation will be straightforward

@TomasTomecek WDYT? didn't we miss anything?

Reply:
lol, I just suggested the same thing :D :D 

I'd call it `archive_root_dir_prefix`: to me, inner_archive suggests that there is an archive inside something

Reply:
@TomasTomecek yes its the output of the previous discussion, and this was just to confirm with you. 

regarding name - to have more flexibility we want to use a template (eg. `blabla-v{version}`)

so final name will be `archive_root_dir_template`

@lachmanfrantisek any objections?